jax/_src/random.py

```diff
@@ -1306,7 +1306,10 @@ def categorical(key: jnp.ndarray,
     _check_shape("categorical", shape, batch_shape)
 
   sample_shape = shape[:len(shape)-len(batch_shape)]
-  return jnp.argmax(gumbel(key, sample_shape + logits.shape, logits.dtype) + logits, axis=axis)
+  return jnp.argmax(
+      gumbel(key, sample_shape + logits.shape, logits.dtype) +
+      jnp.broadcast_to(logits, sample_shape + logits.shape),
```

Comment:
I just realized this is a full `broadcast_to`, which actually materializes a larger array through duplication of the entries. I think we could avoid the rank promotion warning more efficiently by using `lax.expand_dims` instead to insert leading dimension of size 1 up to the desired rank.

Reply:
Done